invenio_cache/lock.py

```diff
@@ -0,0 +1,195 @@
+# -*- coding: utf-8 -*-
+#
+# Copyright (C) 2023 CERN.
+#
+# Invenio-Cache is free software; you can redistribute it and/or modify
+# it under the terms of the MIT License; see LICENSE file for more details.
+"""Lock mechanisms."""
+
+
+from datetime import datetime
+
+from flask import current_app
+
+from invenio_cache.proxies import current_cache
+
+
+class Lock:
+    """Base Lock class."""
+
+    def __init__(self, lock_id):
+        """Initialises the lock instance.
+
+        It does not grant access to the lock itself. To access the lock, call ``lock.acquire``.
+        The lock can be used in two ways:
+
+        1) inside a context:
+
+        .. code-block:: python
+
+            with lock(lock_id) as lock:
+                if lock.acquired():
+                    f()
+
+        2) manually calling each block / unblock mechanism:
+
+        .. code-block:: python
+
+            lock = lock(lock_id)
+            acquired = lock.acquire()
+            if acquired:
+                    f()
+                    lock.release()
+
+        :param lock_id: id of the lock.
+        :type lock_id: str
+        """
+        self.lock_id = lock_id
+        self._acquired = False
+        self._created = datetime.utcnow()
+
+    def __enter__(self):
+        """Entering the context."""
+        return self
+
+    def __exit__(self, exc_type, *args):
+        """Releases the lock if it was previously acquired."""
+        if self.acquired:
+            self.release()
+
+    @property
+    def acquired(self):
+        """Property to check if the lock was acquired."""
+        return self._acquired
+
+    @acquired.setter
+    def acquired(self, value):
+        self._acquired = value
+
+    @property
+    def created(self):
+        """Lock creation date time."""
+        return self._created
+
+    def acquire(self, **kwargs):
+        """Attempts to acquire the lock.
+
+        The method to adcquire the lock must be atomic.
+        It must return a boolean.
+        """
+        raise NotImplementedError
+
+    def release(self):
+        """Attempts to release the lock."""
+        raise NotImplementedError
+
+    def exists(self):
+        """Checks if the lock exists."""
+        raise NotImplementedError
+
+
+class CachedMutex(Lock):
+    """Implements a Mutex using CacheLib API.
+
+    This lock will set a key in a cache when the lock is acquired, and delete the key when
+    the lock is to be deleted.
+    This mechanism is based on the principle that the desired cache's backend provides an atomic
+    write operation (e.g. "write if not exists").
+    """
+
+    _cache = current_cache
+
+    def acquire(self, timeout):
+        """Attempts to acquire the lock.
+
+        The method to adcquire the lock must be atomic.
+
+        :returns: ``True`` if the lock was acquired, ``False`` otherwise .
+        :rtype: boolean
+        :param timeout: lock key timeout.
+        :type timeout: int, optional
+        """
+        success = False
```

Comment:
For all these methods, I think that it is better to succeed or raise a specific exception: if `False` is returned, it does not give info on how to recover.